### PR TITLE
feat(Bonus Pagamenti Digitali): [#176925463] Tracks action to notify the possibility to add a co-badge to the user

### PR DESCRIPTION
**Pull request by debiff**
## Short description
This PR dispatch an action tracked in mixpanel that tell us if a user can add a co-badge card or not.

## List of changes proposed in this pull request
- Check if there is a new card in the `WalletHomeScreen`
- Create the `coBadgeReminder` saga
- Dispatch and track `sendAddCobadgeMessage` to keep track of the possibility of the user.

https://user-images.githubusercontent.com/11773070/113184628-b581f300-9255-11eb-9d05-2132f66ce69b.mov

### CODE REVIEW

File: ts/store/middlewares/analytics.ts [hunk 1]
```diff
@@ -445,6 +446,8 @@ const trackAction = (mp: NonNullable<typeof mixpanel>) => (
         choice: action.payload.choice,
         reason: action.payload.error.message
       });
+    case getType(sendAddCobadgeMessage):
+      return mp.track(action.type, { choice: action.payload });
```
Comment (Undermaken) [inline ts/store/middlewares/analytics.ts:450]: is it really a choice?
```suggestion
      return mp.track(action.type, { canAdd: action.payload });
```

btw what do you think to make the `sendAddCobadgeMessage` an action without payload?
At the end we want to know only these profiles that match this condition. These ones that never sent this action means `choice: false` implicitly

File: ts/sagas/wallet/cobadgeReminder.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  // Extract the cobadgeAbi if there is at least a cobadge card
+  const maybeCobadgeVisibleInWallet: pot.Pot<
+    ReadonlyArray<CreditCardPaymentMethod>,
+    Error
+  > = yield select(cobadgeListVisibleInWalletSelector);
+
+  const cobadgeVisibleInWallet = pot.getOrElse(maybeCobadgeVisibleInWallet, []);
+  const cobadgeAbis = cobadgeVisibleInWallet
+    .filter(c => c.info.issuerAbiCode !== undefined)
```
Comment (Undermaken) [inline ts/sagas/wallet/cobadgeReminder.ts:50]: Is this filter needed?
it is already done by `cobadgeListVisibleInWalletSelector`. Isn't it?

Comment (Undermaken) [reply ts/sagas/wallet/cobadgeReminder.ts:50]: `maybeCobadgeVisibleInWallet`
`cobadgeVisibleInWallet`
`cobadgeAbis`
can be computed just before `enalbedAbis`
If the code returns before that point you have executed few instructions uselessly

File: ts/sagas/wallet/cobadgeReminder.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  if (pot.isSome(maybeCoBadgeAbiConfiguration)) {
+    const coBadgeAbiConfiguration = maybeCoBadgeAbiConfiguration.value;
+
+    // Extract a list of abi that satisfy the following conditions:
+    // - is a bancomat in the wallet of the user
+    // - the abi of the bancomat is in the abiConfiguration list
+    // - the abi of the bancomant has the status enabled in the abiConfiguration list
+    // - there isn't a cobadge card in the wallet of the user with the sami abi
+    const enalbedAbis = bancomatListVisibleInWallet.filter(
```
Comment (Undermaken) [inline ts/sagas/wallet/cobadgeReminder.ts:82]:
```suggestion
    const enabledAbis = bancomatListVisibleInWallet.filter(
```

File: ts/sagas/wallet/cobadgeReminder.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      b =>
+        b.info.issuerAbiCode !== undefined &&
+        coBadgeAbiConfiguration[b.info.issuerAbiCode] !== undefined &&
```
Comment (Undermaken) [inline ts/sagas/wallet/cobadgeReminder.ts:85]:
```suggestion
```
I guess `coBadgeAbiConfiguration[b.info.issuerAbiCode] === StatusEnum.enabled` is just enough

File: ts/sagas/wallet/cobadgeReminder.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        coBadgeAbiConfiguration[b.info.issuerAbiCode] === StatusEnum.enabled &&
+        cobadgeAbis.filter(abi => abi === b.info.issuerAbiCode).length === 0
```
Comment (Undermaken) [inline ts/sagas/wallet/cobadgeReminder.ts:87]:
```suggestion
        !cobadgeAbis.some(abi => abi === b.info.issuerAbiCode)
```

File: ts/sagas/wallet/cobadgeReminder.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    );
+
+    if (enalbedAbis.length > 0) {
+      yield put(sendAddCobadgeMessage(true));
+    } else {
+      yield put(sendAddCobadgeMessage(false));
+    }
```
Comment (Undermaken) [inline ts/sagas/wallet/cobadgeReminder.ts:94]:
```suggestion
yield put(sendAddCobadgeMessage(enalbedAbis.length > 0));
```

Comment (debiff) [reply ts/sagas/wallet/cobadgeReminder.ts]: Thank you for the suggestion, modified in [de22a56](https://github.com/pagopa/io-app/pull/2952/commits/de22a56100ca8a11379a7842aa5704e828dd245c)

File: ts/sagas/wallet/cobadgeReminder.ts (hunk 5 above)
Comment (debiff) [reply ts/sagas/wallet/cobadgeReminder.ts]: Thank you for the suggestion, modified in [de22a56](https://github.com/pagopa/io-app/pull/2952/commits/de22a56100ca8a11379a7842aa5704e828dd245c)

File: ts/sagas/wallet/cobadgeReminder.ts (hunk 4 above)
Comment (debiff) [reply ts/sagas/wallet/cobadgeReminder.ts]: Removed in [d9e85d0](https://github.com/pagopa/io-app/pull/2952/commits/d9e85d0381b3032788703cef991d6579219feab0)

File: ts/sagas/wallet.ts [hunk 7]
```diff
@@ -896,6 +898,9 @@ export function* watchWalletSaga(
     );
   }
 
+  // Check if a user has a bancomat and has not request a cobadge yet and send
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/sagas/wallet.ts]: Fixed in [84d0ff0](https://github.com/pagopa/io-app/pull/2952/commits/84d0ff0e187315e60fee211573048482f8cc0d27)

File: ts/store/actions/wallet/wallets.ts [hunk 8]
```diff
@@ -156,6 +156,14 @@ export const addCreditCardWebViewEnd = createStandardAction(
   "ADD_CREDIT_CARD_WEB_VIEW_END"
 )<AddCreditCardWebViewEndReason>();
 
+export const runSendAddCobadgeMessageSaga = createStandardAction(
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/store/actions/wallet/wallets.ts]: Modified in [7aacb69](https://github.com/pagopa/io-app/pull/2952/commits/7aacb6922bd1a89d50b11a9c3abe4081db61cae0)

File: ts/sagas/wallet/cobadgeReminder.ts (hunk 2 above)
Comment (debiff) [reply ts/sagas/wallet/cobadgeReminder.ts]: For that concern `cobadgeListVisibleInWalletSelector`, yes it's true.

Both modified in [de146d1](https://github.com/pagopa/io-app/pull/2952/commits/de146d1f1d46b18931915c27f001ad8e3b22d722)

File: ts/sagas/wallet/cobadgeReminder.ts (hunk 3 above)
Comment (debiff) [reply ts/sagas/wallet/cobadgeReminder.ts]: Fixed in [715c878](https://github.com/pagopa/io-app/pull/2952/commits/715c87864b660a03adb108b5e9b7446349606f6e)

File: ts/store/middlewares/analytics.ts (hunk 1 above)
Comment (fabriziofff) [reply ts/store/middlewares/analytics.ts]: In order to implement the required feature we need the payload, to distinguish and avoid sending the message to those who add a BANCOMAT and then shortly after also a cobadge card.
eg:
1) User add a BANCOMAT candidate to the addition of a cobadge card, we send the action `canAdd: true`.
2) The user add a co-badge card.
3) We send the notification message

if in step two we do not update the user's information, we will send the message even if the user has already inserted the card

**Review by fabriziofff — COMMENTED** [review]
@debiff could you add the new tracked action with the payload to https://docs.google.com/spreadsheets/d/1bXNRcTDKU-5l2RqCcwSbT7Y1M9QQTWtS_f-3-AIqSYk/edit#gid=1909779257&range=A89:D89 ?

**Review by fabriziofff — CHANGES_REQUESTED** [review]

File: ts/store/middlewares/analytics.ts [hunk 9]
```diff
@@ -447,6 +448,8 @@ const trackAction = (mp: NonNullable<typeof mixpanel>) => (
         choice: action.payload.choice,
         reason: action.payload.error.message
       });
+    case getType(sendAddCobadgeMessage):
+      return mp.track(action.type, { canAdd: action.payload });
```
[the remark this answers is not in the document]
Comment (debiff) [reply ts/store/middlewares/analytics.ts]: Sure, moved in [19cdcf0](https://github.com/pagopa/io-app/pull/2952/commits/19cdcf08e416cb9102477ee96406c5f302247bf5)

**Review by fabriziofff — APPROVED** [review]

**Review by Undermaken — APPROVED** [review]

**Pull request merged**